src/transformers/modeling_encoder_decoder.py

```diff
@@ -16,37 +16,65 @@
 
 
 import logging
-import os
-
-from torch import nn
 
 from .modeling_auto import AutoModel, AutoModelWithLMHead
+from .modeling_utils import PreTrainedModel
+from .configuration_encoder_decoder import EncoderDecoderConfig
 
 
 logger = logging.getLogger(__name__)
 
 
-class PreTrainedEncoderDecoder(nn.Module):
+class EncoderDecoderModel(PreTrainedModel):
     r"""
-        :class:`~transformers.PreTrainedEncoderDecoder` is a generic model class that will be
+        :class:`~transformers.EncoderDecoder` is a generic model class that will be
         instantiated as a transformer architecture with one of the base model
-        classes of the library as encoder and (optionally) another one as
+        classes of the library as encoder and another one as
         decoder when created with the `AutoModel.from_pretrained(pretrained_model_name_or_path)`
-        class method.
+        class method for the encoder and `AutoModelWithLMHead.from_pretrained(pretrained_model_name_or_path)` class method for the decoder.
     """
+    config_class = EncoderDecoderConfig
+
+    def __init__(self, config=None, encoder=None, decoder=None):
+        assert config is not None or (encoder is not None and decoder is not None), "Either a configuration or an Encoder and a decoder has to be provided"
+        if config is None:
+            config = EncoderDecoderConfig.from_encoder_decoder_config(encoder.config, decoder.config)
+        else:
+            assert isinstance(config, self.config_class), "config: {} has to be of type {}".format(config, self.config_class)
+        # initialize with config
+        super().__init__(config)
+
+        if encoder is None:
+            encoder = AutoModel.from_config(config.encoder)
+
+        if decoder is None:
+            decoder = AutoModelWithLMHead.from_config(config.decoder)
 
-    def __init__(self, encoder, decoder):
-        super().__init__()
         self.encoder = encoder
         self.decoder = decoder
+        assert (
+            self.encoder.get_output_embeddings() is None
+        ), "The encoder {} should not have a LM Head. Please use a model without LM Head"
+
+    def tie_weights(self):
+        # for now no weights tying in encoder-decoder
+        pass
+
+    def get_encoder(self):
+        return self.encoder
+
+    def get_decoder(self):
+        return self.decoder
+
+    def get_input_embeddings(self):
+        return self.encoder.get_input_embeddings()
+
+    def get_output_embeddings(self):
+        return self.decoder.get_output_embeddings()
 
     @classmethod
-    def from_pretrained(
-        cls,
-        encoder_pretrained_model_name_or_path=None,
-        decoder_pretrained_model_name_or_path=None,
-        *model_args,
-        **kwargs
+    def from_encoder_decoder_pretrained(
```

Comment:
@thomwolf  As @sshleifer suggested I changed the name here to be consistent with the config name. In order to load a "complete" encoder-decoder model the normal `from_pretrained` fn can be used. 

Note that `EncoderDecoderModel` can be loaded via both:

`from_encoder_decoder_pretrained` and `from_pretrained` (as is the case for the config)

Reply:
Ok!